lib/matplotlib/backends/backend_nbagg.py

```diff
@@ -229,76 +250,46 @@ def closer(event):
     """
-    def __init__(self, manager):
-        self.supports_binary = None
-        self.manager = manager
-        self.uuid = str(uuid())
-        # Publish an output area with a unique ID. The javascript can then
-        # hook into this area.
-        display(HTML("<div id=%r></div>" % self.uuid))
-        try:
-            self.comm = Comm('matplotlib', data={'id': self.uuid})
-        except AttributeError:
-            raise RuntimeError('Unable to create an IPython notebook Comm '
-                               'instance. Are you in the IPython notebook?')
-        self.comm.on_msg(self.on_message)
-
-        manager = self.manager
-        self._ext_close = False
-
-        def _on_close(close_message):
-            self._ext_close = True
-            manager.remove_comm(close_message['content']['comm_id'])
-            manager.clearup_closed()
-
-        self.comm.on_close(_on_close)
-
-    def is_open(self):
-        return not (self._ext_close or self.comm._closed)
-
-    def on_close(self):
-        # When the socket is closed, deregister the websocket with
-        # the FigureManager.
-        if self.is_open():
-            try:
-                self.comm.close()
-            except KeyError:
-                # apparently already cleaned it up?
-                pass
-
-    def send_json(self, content):
-        self.comm.send({'data': json.dumps(content)})
-
-    def send_binary(self, blob):
-        # The comm is ascii, so we always send the image in base64
-        # encoded data URL form.
-        data = b64encode(blob)
-        if six.PY3:
-            data = data.decode('ascii')
-        data_uri = "data:image/png;base64,{0}".format(data)
-        self.comm.send({'data': data_uri})
-
-    def on_message(self, message):
-        # The 'supports_binary' message is relevant to the
-        # websocket itself.  The other messages get passed along
-        # to matplotlib as-is.
-
-        # Every message has a "type" and a "figure_id".
-        message = json.loads(message['content']['data'])
-        if message['type'] == 'closing':
-            self.on_close()
-            self.manager.clearup_closed()
-        elif message['type'] == 'supports_binary':
-            self.supports_binary = message['value']
-        else:
-            self.manager.handle_json(message)
+    if (check_nbextension('matplotlib') or
+            check_nbextension('matplotlib', True)):
+        return
+
+    # Make a temporary directory so we can wrap mpl.js in a requirejs define().
+    tempdir = mkdtemp()
+    path = os.path.join(os.path.dirname(__file__), "web_backend")
+    shutil.copy2(os.path.join(path, "nbagg_mpl.js"), tempdir)
+
+    with open(os.path.join(path, 'mpl.js')) as fid:
+        contents = fid.read()
+
+    with open(os.path.join(tempdir, 'mpl.js'), 'w') as fid:
+        fid.write('define(["jquery"], function($) {\n')
+        fid.write(contents)
+        fid.write('\nreturn mpl;\n});')
```

Comment:
The current method is to inject the javascript directly into the notebook.  The preferred method for IPython third-party widgets is to store the javascript as an nbextension that is loaded by the notebook server, and it does not get stored with the notebook document.  This is served from a central location:

```bash
$ ls /Users/ssilvester/Library/Jupyter/nbextensions/
bqplot     matplotlib qgridjs
```